### PR TITLE
HDDS-5112. Recon Insights page does not list buckets when only one volume is present.

### PR DESCRIPTION
**Pull request by smitajoshi12**
What changes were proposed in this pull request?
When there is only one volume present in the cluster, the buckets dropdown does not list is getting displayed.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5112

## How was this patch tested?
Manually

image

### CODE REVIEW

**Comment by kerneltime** [comment]
@GeorgeJahad @SaketaChalamchala

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 1]
```diff
@@ -89,8 +89,15 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
     // selected buckets value should be reset to all buckets
     let selectedBuckets = [allBucketsOption];
     // Update bucket options only if one volume is selected
-    if (selectedVolumes && selectedVolumes.length === 1) {
-      const selectedVolume = selectedVolumes[0].value;
+    if (selectedVolumes && selectedVolumes.length !== null && (selectedVolumes.length === 2 && selectedVolumes[0].value == '*') || (selectedVolumes.length === 1)){
```
Comment (DaveTeng0) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: just curious, what does the * mean? and why it's the first element in selectedVolumes?

Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: const allVolumesOption: IOption = {
  label: 'All Volumes',
  value: '*'
};

When we are selecting only one Volume then All volumes is getting selected by Default it is natural behaviour of dropdown so we are checking condition for One volume and multiple volume.

Comment (DaveTeng0) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: i see!! Thanks!

**Review by dombizita — COMMENTED** [review]
Thanks for the patch @smitajoshi12, I tried it and it works as expected, there is one small comment inline. Beside that it looks good to me.

**Comment by dombizita** [comment]
Also I just checked and the `build-branch / build (8)` in the CI failed with this message:
```Error:  Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.12.0:npx (Build frontend) on project ozone-recon: Failed to run task: 'npx pnpm@5.18.9 run build' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]```
Can you take a look at it? @smitajoshi12

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 2]
```diff
@@ -89,8 +89,15 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
...
+      let selectedVolume;
+      if (selectedVolumes.length === 1)
+      {
+      selectedVolume = selectedVolumes[0].value;
+      }
+      else {
+      selectedVolume = selectedVolumes[1].value;
+      }
```
[the remark this answers is not in the document]
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: Indentation Issue  Resolved.

**Comment by smitajoshi12** [comment]
> Also I just checked and the `build-branch / build (8)` in the CI failed with this message: `Error: Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.12.0:npx (Build frontend) on project ozone-recon: Failed to run task: 'npx pnpm@5.18.9 run build' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]` Can you take a look at it? @smitajoshi12


@dombizita  Hi Zita  Now I can see build is success.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx [hunk 3]
```diff
@@ -89,8 +89,14 @@ export class Insights extends React.Component<Record<string, object>, IInsightsS
     // selected buckets value should be reset to all buckets
     let selectedBuckets = [allBucketsOption];
     // Update bucket options only if one volume is selected
-    if (selectedVolumes && selectedVolumes.length === 1) {
-      const selectedVolume = selectedVolumes[0].value;
+    if (((selectedVolumes && selectedVolumes.length != null) && (selectedVolumes.length === 2 && selectedVolumes[0].value === '*')) || (selectedVolumes.length === 1)){
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: Is the `selectedVolumes.length != null` necessary? Later we are checking if the size is either 1 or 2.
```suggestion
    if ((selectedVolumes && (selectedVolumes.length === 2 && selectedVolumes[0].value === '*')) || (selectedVolumes.length === 1)){
```

Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: Yes Sometimes if length and selectedVolumes[0].value === '*' are both mandatory condition.

Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx:92]: I think the brackets wasn't at good place previously. So there is a `AND` after the `selectedVolumes`(like it was originally in the code) and after that inside the brackets there is an `OR`, where we need the length to be either 1 or 2. If we are checking that it either needs to be 1 or 2, why do we need the null check before that? Please check my suggestion in this comment.
```suggestion
    if (selectedVolumes && ((selectedVolumes.length === 2 && selectedVolumes[0].value === '*') || (selectedVolumes.length === 1))){
```

Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/insights/insights.tsx]: Done Changes Suggested by Zita

**Review by smengcl — APPROVED** [review]

**Comment by smengcl** [comment]
Thanks @smitajoshi12 for the patch. Thanks @dombizita for the review.

**Pull request merged**